presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/rtree/Flatbush.java

```diff
@@ -0,0 +1,306 @@
+ * object's Hilbert curve index.  Hilbert curves are naturally hierarchical, so
+ * successively grouping children and their parents will give a naturally nested
+ * structure.  This means we only need to sort the items once.
+ * <p>
+ * If sort time is a problem, we actually just need to "partition" into groups
+ * of `degree`, since the order of the children of a single parent doesn't
+ * matter.  This could be done with quicksort, stopping once all items at index
+ * `n * degree` are correctly placed.
+ * <p>
+ * Original implementation in Javascript: https://github.com/mourner/flatbush
+ */
+public class Flatbush<T extends HasExtent>
+{
+    // Number of coordinates to define an envelope
+    @VisibleForTesting
+    static final int ENVELOPE_SIZE = 4;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(Flatbush.class).instanceSize();
+    private static final int DEFAULT_DEGREE = 16;
+    // Number of children per node
+    private final int degree;
+    // Offsets in tree for each level
+    private final int[] levelOffsets;
+    // Each node has four doubles: xMin, yMin, xMax, yMax
+    private final DoubleBigArray tree;
+    private final T[] items;
+
+    /**
+     * Build Flatbush RTree for `items`.
+     *
+     * @param items Items to index.
+     * @param degree Number of children for each intermediate node.
+     */
+    public Flatbush(T[] items, int degree)
+    {
+        checkArgument(degree > 0, "degree must be positive");
+        this.degree = degree;
+        this.items = requireNonNull(items, "items is null");
+        this.levelOffsets = calculateLevelOffsets(items.length, degree);
+        this.tree = buildTree();
+    }
+
+    /**
+     * Build Flatbush RTree for `items` with default number of children per node.
+     * <p>
+     * This will use the default degree.
+     *
+     * @param items Items to index.
+     */
+    public Flatbush(T[] items)
+    {
+        this(items, DEFAULT_DEGREE);
+    }
+
+    /**
+     * Calculate the indices for each level.
+     *
+     * A given level contains a certain number of items.  We give it a capacity
+     * equal to the next multiple of `degree`, so that each parent will have
+     * an equal number (`degree`) of children.  This means the next level will
+     * have `capacity/degree` nodes, which yields a capacity equal to the next
+     * multiple, and so on, until the number of items is 1.
+     *
+     * Since we are storing each node as 4 doubles, we will actually multiply
+     * every index by 4.
+     */
+    private static int[] calculateLevelOffsets(int numItems, int degree)
+    {
+        List<Integer> offsets = new ArrayList<>();
+        // Leaf nodes start at 0, root is the last element.
+        offsets.add(0);
+        int level = 0;
+        while (numItems > 1) {
+            // The number of children will be the smallest multiple of degree >= numItems
+            int numChildren = (int) Math.ceil(1.0 * numItems / degree) * degree;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).